runtime/lua/vim/_memoize.lua

```diff
@@ -0,0 +1,42 @@
+--- Module for private utility functions
+
+local M = {}
+
+local function concat_hash(...)
+  return table.concat({...})
+end
+
+local function idx_hash(idx)
+  return function(...)
+    return select(idx, ...)
+  end
+end
+
+--- @generic F: function
+--- @param fn F
+--- @param hash_fn? integer|'concat'|fun(...): any Hash function to to create a hash to
+---                   stored results. When an integer, refers to the argument
+---                   to use as a hash. Defaults to 1.
+--- @return F
+return function(fn, hash_fn)
+  ---@type table<any,table<any,any>>
+  local cache = setmetatable({}, { __mode = 'kv' })
+
+  hash_fn = hash_fn or 1
+
+  if type(hash_fn) == 'number' then
+    hash_fn = idx_hash(hash_fn)
+  elseif hash_fn == 'concat' then
+    hash_fn = concat_hash
+  end
+
+  return function(...)
+    local key = hash_fn(...)
+    if cache[key] == nil then
+      cache[key] = vim.F.pack_len(fn(...))
+    end
+
+    local v = vim.deepcopy(cache[key])
```

Comment:
Is it a must to deep copy the result of the cache, can't we reuse the result directly

Reply:
Well yeah, you don't want to risk mutating the cached values.

Reply:
> Well yeah, you don't want to risk mutating the cached values.

Then having a way not to copy it might be nice. I can see some query handling cases where I don't mutate the cached values, but don't want to repeatedly deepcopying a large table

Reply:
I'm not the expert here but I suspect that returning copies could mess up with the weak map.

Reply:
How come? The copies are unrelated to each other.

If the function is only returning strings|numbers|booleans|functions, then the copying should have no effect.

Reply:
Does this handle hash collision?

Reply:
No. The hash is just determined for a key. If there's a collision (via the same args or different) then the memoized value is returned.

This means you could use just the e.g. the buffer number as a key, if you know other arguments won't have an effect.

Reply:
I think if it doesn't then it's a huge footgun since the default of `hash_fn` is `1` which is general not unique. Actually `"concat"` is not unique either since you can have "foo" + "bar" = "foob" + "ar".

So if we are not going to do a full hash table (array at each hash) we should probably add a big warning "HASH MUST BE UNIQUE" and adjust usages accordingly

Reply:
sorry, not exactly that message since, as you noted, there is a use-case for non-unique hash, but something more nuanced that still brightly signals the behavior

Reply:
> How come? The copies are unrelated to each other.
> 
> If the function is only returning strings|numbers|booleans|functions, then the copying should have no effect.

Yeah my comment was a reply to @lucario387 about not returning copies.

Reply:
for the context of treesitter, the hash will mostly be unique. as `TSNode:id()` is unique, so even if we get two similar s-expr, the root node's id will still be different, leading to two different cache keys

Reply:
are `TSNode` immutable or can you change a child of the Node and still have the same parent ID? Because that should invalidate the query

if we were really bikeshedding, we could add an invalidation API (though i suspect this PR might be most immediately useful for things with infinite TTL such as looking up files, so it wouldn't need it)

Reply:
> Is it a must to deep copy the result of the cache

There's another issue with the deepcopy actually:

```
vim/shared.lua:0: Cannot deepcopy object of type userdata
```

Reply:
Good catch. I'll add a comment.

Reply:
Decided to remove `deepcopy()` instead.